content-services/content-sources-backend
# Refs 533: Add repository introspection endpoint

## Summary
This will also have a frontend PR, but I'll put this up as a draft in the meantime for judgement.

- Adds `/repository/{uuid}/introspect` endpoint to introspect a repository associated with the given repository configuration UUID.
- Adds `failed_introspection_count` to the `repositories` table, which represents the number of consecutive failed introspections and is used to skip introspection if a certain limit is reached.
- Adds a `reset_count` boolean param to this endpoint that will reset the `failed_introspections_count` column back to zero.
- Changed format for timestamps returned by API to be RFC3339.
## Testing steps
1. Migrate DB
2. Set the `OPTIONS_INTROSPECT_API_TIME_LIMIT_SEC` environment variable to 10. This should make it should you cannot introspect a repository more than once every 10 seconds. You should also test setting this  option in config.yaml, `options.introspect_api_time_limit_sec`.
3. Introspect an invalid repository with the endpoint, without passing the reset_count param. The failed introspections count should increase. You can verify this by fetching the repository from its GET endpoint.
4. Introspect an invalid repository and pass the reset_count param. The failed introspections count should become 1 after introspection fails.
6. Introspect a valid repository that was failing previously. The failed introspections count should reset to 0.
7. Introspect a repository over the FailedIntrospectionsLimit (20), using the CLI, without using the `--force` flag. It should be skipped. Doing that again, with the `--force` flag, should return an error.
